### PR TITLE
ci: add support for Nvme/TLS connection

### PR DESCRIPTION
**Pull request by glimchb**
Signed-off-by: Boris Glimcher <Boris.Glimcher@emc.com>

### CODE REVIEW

**Comment by glimchb** [comment]
waiting on https://github.com/opiproject/opi-api/pull/308

File: pkg/frontend/nvme_controller.go [hunk 1]
```diff
@@ -75,15 +75,22 @@ func NewNvmeTCPTransport(listenAddr string) NvmeTransport {
 	}
 }
 
-func (c *nvmeTCPTransport) Params(_ *pb.NvmeController, nqn string) (spdk.NvmfSubsystemAddListenerParams, error) {
+func (c *nvmeTCPTransport) Params(ctrl *pb.NvmeController, nqn string) (spdk.NvmfSubsystemAddListenerParams, error) {
 	result := spdk.NvmfSubsystemAddListenerParams{}
 	result.Nqn = nqn
-	result.SecureChannel = false
+
```
Comment (glimchb) [inline pkg/frontend/nvme_controller.go:81]: hack
should be addressed in https://github.com/opiproject/opi-api/pull/375

File: docker-compose.yml [hunk 2]
```diff
@@ -66,7 +66,7 @@ services:
         condition: service_healthy
       redis:
         condition: service_healthy
-    command: sh -c "/opi-spdk-bridge -grpc_port=50051 -http_port=8082 -spdk_addr=/var/tmp/spdk.sock -tcp_trid=$$(getent hosts spdk | awk '{ print $$1 }'):7777"
+    command: sh -c "/opi-spdk-bridge -grpc_port=50051 -http_port=8082 -spdk_addr=/var/tmp/spdk.sock -tcp_trid=$$(getent hosts spdk | awk '{ print $$1 }'):777"
```
Comment (glimchb) [inline docker-compose.yml:69]: ugly hack! should be addressed in https://github.com/opiproject/opi-api/pull/375

File: pkg/frontend/nvme_controller.go [hunk 3]
```diff
@@ -75,15 +74,22 @@ func NewNvmeTCPTransport(listenAddr string) NvmeTransport {
...
+	if *ctrl.Spec.NvmeControllerId < 0 {
+		result.SecureChannel = true
+	} else {
+		result.SecureChannel = false
+	}
+
```
Comment (glimchb) [inline pkg/frontend/nvme_controller.go:86]: ugly hack! should be addressed in https://github.com/opiproject/opi-api/pull/375

File: pkg/frontend/nvme_controller.go [hunk 4]
```diff
@@ -129,11 +138,27 @@ func (s *Server) CreateNvmeController(_ context.Context, in *pb.CreateNvmeContro
 		msg := fmt.Sprintf("Could not create CTRL: %s", in.NvmeController.Name)
 		return nil, status.Errorf(codes.InvalidArgument, msg)
 	}
+
+	// TODO: this is hard coded to specific host
+	params2 := spdk.NvmfSubsystemAddHostParams{
```
Comment (glimchb) [inline pkg/frontend/nvme_controller.go:143]: this is needed since we switched to `AllowAnyHost: false` above

File: docker-compose.yml [hunk 5]
```diff
@@ -16,6 +16,7 @@ services:
       - "9009:9009"
```
Comment (glimchb) [inline docker-compose.yml:16]: document ports better:
- 4444 our default spdk no tls before any RPC is available (compose setting for sanity check)
- 5555 our default spdk with tls before any RPC is available (compose setting for sanity check)
- 7777 from cmdline arguments when starting bridge, should be replaced by API change https://github.com/opiproject/opi-api/pull/375

File: pkg/backend/nvme_path.go [hunk 6]
```diff
@@ -295,7 +295,7 @@ func (s *Server) numberOfPathsForController(controllerName string) int {
 }
 
 func (s *Server) keyToTemporaryFile(pskKey []byte) (string, error) {
-	keyFile, err := s.psk.createTempFile("", "opikey")
+	keyFile, err := s.psk.createTempFile("/var/tmp", "opikey")
```
Comment (glimchb) [inline pkg/backend/nvme_path.go:298]: this is fix for #677 issue

File: pkg/frontend/nvme_controller.go [hunk 7]
```diff
@@ -75,15 +74,25 @@ func NewNvmeTCPTransport(listenAddr string) NvmeTransport {
...
+	} else {
+		result.SecureChannel = false
+		port = *ctrl.Spec.NvmeControllerId
+	}
+
 	result.ListenAddress.Trtype = "tcp"
 	result.ListenAddress.Traddr = c.listenAddr.String()
-	result.ListenAddress.Trsvcid = c.listenPort
+	result.ListenAddress.Trsvcid = fmt.Sprintf("%s%d", c.listenPort, port)
 	result.ListenAddress.Adrfam = c.protocol
```
Comment (glimchb) [inline pkg/frontend/nvme_controller.go:93]: Ugly hack! maybe introduce new cli flag for TLS trid ? and use it here instead of attaching ctrl id to port?

File: pkg/frontend/nvme_controller.go (hunk 4 above)
Comment (artek-koltun) [reply pkg/frontend/nvme_controller.go:143]: I think we could extend opiproject/opi-api#375 and wrap `FabricsEndpoint` into a `FabricsPath` containing also host nqn and psk
```proto
oneof path {
  PciPath pcie_id = 3 [(google.api.field_behavior) = OPTIONAL];
  FabricsPath fabrics_id = 4 [(google.api.field_behavior) = OPTIONAL];
}
...
message FabricsPath {
  FabricsEndpoint endpoint = 1 [(google.api.field_behavior) = REQUIRED];
  string host_nqn = 2 [(google.api.field_behavior) = REQUIRED]; // we can make it OPTIONAL later if needed. It won't be a breaking change
  bytes psk = 3 [(google.api.field_behavior) = OPTIONAL];
}

message PciePath {
  PcieEndpoint endpoint = 1 [(google.api.field_behavior) = REQUIRED];
}
```
Then we need to rework `nvmeTCPTransport` in either way:
1. It returns a list of cmds with parameters to execute for Create and Delete
``` go
type spdkCmd struct {
  callName string
  params any
}
...
func (c *nvmeTCPTransport) CreateParams(ctrlr *pb.NvmeController, nqn string) ([]spdkCmd, error) {
  result := []spdkCmd{}
  // add nvmf_subsystem_add_listener cmd to result
  // add nvmf_subsystem_add_host cmd to result
  if len(ctrlr. ... .Psk) != 0 {
    // add secure channel to nvmf_subsystem_add_listener
    // set psk to nvmf_subsystem_add_host 
  }
  return result, nil
}
func (c *nvmeTCPTransport) DeleteParams(ctrlr *pb.NvmeController, nqn string) (spdkCmd, error) {
  return /* nvmf_subsystem_remove_listener cmd*/, nil
}
```
and in `CreateNvmeController` execute returned cmds one by one

2. Another option is to move SPDK calls into NvmeTransports instead of returning parameters
```go
type NvmeTransport interface {
    // Params(ctrlr *pb.NvmeController, nqn string) (spdk.NvmfSubsystemAddListenerParams, error)
	CreateController(ctrlr *pb.NvmeController, nqn string) error
    DeleteController(ctrlr *pb.NvmeController, nqn string) error
}
```

File: scripts/tests.sh [hunk 8]
```diff
@@ -31,6 +32,7 @@ fi
 
 # Check exported port also works (host network)
 docker run --network=host --rm docker.io/namely/grpc-cli ls 127.0.0.1:50051
+docker run --network=host --rm docker.io/curlimages/curl:8.3.0 curl -qkL http://127.0.0.1:8082/v1/inventory/1/inventory/2
```
Comment (artek-koltun) [inline scripts/tests.sh:35]: probably it should be in separate PR

File: docker-compose.yml (hunk 2 above)
Comment (artek-koltun) [reply docker-compose.yml:70]: or we can pass `-tcp_trid="not_secured_endpoint;secured_endpoint"` since at this stage tcp transport in frontend is more for demo purposes and will be removed after https://github.com/opiproject/opi-api/pull/375 is implemented

File: pkg/frontend/nvme_controller.go (hunk 7 above)
Comment (glimchb) [reply pkg/frontend/nvme_controller.go]: went in different direction

File: pkg/frontend/nvme_subsystem.go [hunk 9]
```diff
@@ -80,6 +80,7 @@ func (s *Server) CreateNvmeSubsystem(_ context.Context, in *pb.CreateNvmeSubsyst
 		params := spdk.NvmfSubsystemAddHostParams{
 			Nqn:  in.NvmeSubsystem.Spec.Nqn,
 			Host: in.NvmeSubsystem.Spec.Hostnqn,
+			Psk:  "/tmp/opikey.txt",
```
Comment (glimchb) [inline pkg/frontend/nvme_subsystem.go:83]: I think to merge like this and then move PSK file save into utils and use it here... WDYT @artek-koltun ?

**Review by artek-koltun — APPROVED** [review]
LGTM!

File: pkg/frontend/nvme_subsystem.go (hunk 9 above)
Comment (artek-koltun) [reply pkg/frontend/nvme_subsystem.go:83]: agree
